### PR TITLE
[2D] Enable 2D DTensor state_dict for FSDP + TP

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110925
* __->__ #110846
* #110831

This PR adds a `chunk_dtensor()` method to fsdp/_fsdp_extensions.py and the actual implementation of `chunk_dtensor()` in tensor/parallel/fsdp.py. This enables FSDP to return 2D DTensor state_dict when composing FSDP with TP. 

cc. @fegin

### CODE REVIEW

File: torch/distributed/tensor/parallel/fsdp.py [hunk 1]
```diff
@@ -220,6 +221,66 @@ def _chunk_tensor(
         )
 
 
+def _chunk_dtensor(
```
Comment (fduwjj) [inline torch/distributed/tensor/parallel/fsdp.py:224]: Can we reuse `distribute_tensor` API?

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 2]
```diff
@@ -401,6 +401,57 @@ def test_2d_e2e_training_use_orig_params(self):
     def test_2d_e2e_training_not_use_orig_params(self):
         self._test_2d_e2e_training(recompute_activation=True)
 
+    @with_comms
+    @skip_if_lt_x_gpu(4)
+    def test_2d_state_dict(self):
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:406]: Nice!!!

File: torch/distributed/tensor/parallel/fsdp.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/tensor/parallel/fsdp.py:224]: Do you mean the DTensor.from_local().redistribute() part? Since we already know the layout here as well, so for now, it's better to use DTensor.from_local().redistribute().

**Review by fegin — APPROVED** [review]
LGTM, left some nits.

File: torch/distributed/tensor/parallel/fsdp.py [hunk 3]
```diff
@@ -220,6 +221,66 @@ def _chunk_tensor(
...
+    corresponding chunk as the local tensor to create a DTensor.
+    """
+    parent_mesh = mesh_resources.get_parent_mesh(device_mesh)
+    if parent_mesh is None:
+        raise RuntimeError("No parent device_mesh is found for FSDP device_mesh.")
+    if parent_mesh.ndim != 2:
+        raise RuntimeError(
+            f"Found parent device_mesh of ndim={parent_mesh.ndim},",
+            "but only 2D meshes are currently supported.",
+        )
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/tensor/parallel/fsdp.py:240]: Good point. Maybe we should add a check in TP so that would happen during initialization so that would be earlier than state_dict call. For example, during TP ctor, if TP has a parent mesh and the dim idx of TP is not the innermost dimenstion, we error out say we don't support the composition. Wdyt?

File: torch/distributed/tensor/parallel/fsdp.py [hunk 4]
```diff
@@ -220,6 +221,66 @@ def _chunk_tensor(
...
+
+    # We need to explicitly call .detach() to return a new tensor detached from the current graph.
+    tensor = tensor.clone().detach()
+
+    # if a tensor has not yet sharded by TP
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/tensor/parallel/fsdp.py:245]: If we have a model with three layers and we use PairwiseParallel() for TP, the 1st and 2nd layer will be the pair and leave the 3rd layer as it is. In this case, TP would not shard the third layer. 

Similarly, when we have a model with only one layer and we use PairwiseParallel(), it is basically a no-op, since TP doesn't do any sharding to the model.

[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/tensor/parallel/fsdp.py:245]: I meant should we add the detail you described in the the comment so that we won't get lost in the future?

File: torch/distributed/tensor/parallel/fsdp.py [hunk 5]
```diff
@@ -220,6 +221,66 @@ def _chunk_tensor(
...
+        raise RuntimeError("No parent device_mesh is found for FSDP device_mesh.")
+    if parent_mesh.ndim != 2:
+        raise RuntimeError(
+            f"Found parent device_mesh of ndim={parent_mesh.ndim},",
+            "but only 2D meshes are currently supported.",
+        )
+
+    # We need to explicitly call .detach() to return a new tensor detached from the current graph.
+    tensor = tensor.clone().detach()
+
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/fsdp.py:244]: not sure why we need the clone/detach as it's not autograd related?

**Review by wanchaol — APPROVED** [review]
Nice!

File: torch/distributed/tensor/parallel/fsdp.py (hunk 5 above)
Comment (wz337) [reply torch/distributed/tensor/parallel/fsdp.py:244]: > not sure why we need the clone/detach as it's not autograd related?

We need .clone() for sure. From my understanding, the tensor is a view into the FlatParam. If the FlatParam changes, the view will also be changed. The .detach() might not be needed as it should be under @no_grad context from state_dict.  cc. @fegin Correct me if I am wrong.

File: torch/distributed/tensor/parallel/fsdp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/tensor/parallel/fsdp.py:240]: Added composition validation logic during TP initialization in https://github.com/pytorch/pytorch/pull/111001
